firewood/src/merkle/node.rs

```diff
@@ -824,3 +670,182 @@ impl Storable for Node {
         }
     }
 }
+
+pub trait NodeCodec {
+    fn encode<S: ShaleStore<Node>>(node: &NodeType, store: &S) -> Vec<u8>;
+
+    fn decode(buf: &[u8]) -> Result<NodeType, Error>;
+}
+
+pub struct BinCodec;
+
+impl NodeCodec for BinCodec {
+    fn encode<S: ShaleStore<Node>>(node: &NodeType, store: &S) -> Vec<u8> {
+        match node {
+            NodeType::Leaf(n) => BinCodec::encode_leaf(n),
+            NodeType::Extension(n) => BinCodec::encode_ext(n, store),
+            NodeType::Branch(n) => BinCodec::encode_branch(n, store),
```

Comment:
🤔 

Something is off here, but it's not immediately obvious what it is. 

Using a ZST (Zero-Sized Type) to specify the encoding type is the right way to do things as you've done... But I think the way that you actually want to do this is more like the following:
```rust
struct Merke {
    // ...
    format: PhantomData<EncodingFormat>,
}

// shortcut so we don't have to write this out every time
type Store = ShaleStore<Node>;

// could come up with a better name here
pub trait Encode<EncodingFormat> {
    type Error;
    
    fn encode<S: Store>(self, store: S) -> Result<Vec<u8>, Self::Error>;
    
    fn decode(bytes: &[u8]) -> Result<Self, Self::Error>;
}
```
Then implement `Encode` for `NodeType`, `ExtNode`, `LeafNode`, and `BranchNode`, where `NodeType` just delegates to the inner implementation. 

I'm not sure if you need to bind `EncodingFormat` with a marker-trait. If the compiler complains, you could call the marker-trait `EncodingFormat` and change the generic argument to `E: EncodingFormat`, but *I don't think* it'll complain. 
    
However, I could be wrong about where to place the outermost `EncodingFormat` generic, but reviewing the code in this PR seems to point me toward the tree itself. 

 I think you will have to further parameterize `get_encoded` as well, but I could be wrong. When you are done with this PR, there should be no way to serialize with one `Codec` and deserialize with another.

Reply:
Further thoughts (that I'm not completely sure of):
There may be a better way to do all of this with something like the following:
```rust
pub trait BinaryCodec {
    // depending on the Serializer and Deserializer implementations, these will likely always be the same, but should not be required to be the same
    type SerializeError;
    type DeserializeError = Self::SerializeError;

    fn serialize<T: Serialize>(&self, t: &T) -> Result<T, Self::SerializeError>;
    
    fn deserialize<'de, T: Deserialize<'de>>(&self, bytes: &[u8]) -> Result<T, Self::DeserializeError>;
}

#[derive(Default)]
struct Bincode(bincode::DefaultOptions); 

impl BinaryCodec for Bincode {
    type SerializeError: bincode::Error;

    fn serialize<T: Serialize>(&self, t: &T) -> Result<T, Self::SerializeError> {
        self.0.clone().serialize(t)
    }
    
    fn deserialize<'de, T: Deserialize<'de>>(&self, bytes: &[u8]) -> Result<T, Self::DeserializeError> {
        self.0.clone().deserialize(bytes)
    }
}
```
Or something along those lines. Then, our node types can simply implement `Serialize` and `Deserialize` and the `Merkle` type would be generic over the `BinaryCodec`. It's likely that we would want to attach a `&'static Bincode` to the database itself to avoid copying the config on every new tree (view).